packages/modelence/src/migration/index.ts

```diff
@@ -1,35 +1,97 @@
+import { acquireLock, releaseLock } from '@/lock';
 import { Module } from '../app/module';
 import { dbMigrations } from './db';
+import { logInfo } from '../telemetry';
 
 export type MigrationScript = {
   version: number;
   description: string;
-  handler: () => Promise<void>;
+  handler: () => Promise<string | void>;
 };
 
 export async function runMigrations(migrations: MigrationScript[]) {
   if (migrations.length === 0) {
     return;
   }
 
+  const hasLock = await acquireLock('migrations');
+
+  if (!hasLock) {
+    logInfo('Another instance is running migrations. Skipping migration run.', {
+      source: 'migrations',
+    });
+    return;
+  }
+
   const versions = migrations.map(({ version }) => version);
 
-  const existingVersions = await dbMigrations.fetch({ version: { $in: versions } });
+  const existingVersions = await dbMigrations.fetch({
+    version: { $in: versions },
+    status: { $ne: 'failed' },
+  });
   const existingVersionSet = new Set(existingVersions.map(({ version }) => version));
   const pendingMigrations = migrations.filter(({ version }) => !existingVersionSet.has(version));
 
   if (pendingMigrations.length === 0) {
     return;
   }
 
-  console.log(`Running migrations (${pendingMigrations.length})...`);
+  logInfo(`Running migrations (${pendingMigrations.length})...`, {
+    source: 'migrations',
+  });
   for (const { version, description, handler } of pendingMigrations) {
-    console.log(`Running migration v${version}: ${description}`);
-    // TODO: adjust to handle multiple containers and race conditions
-    await dbMigrations.insertOne({ version, appliedAt: new Date() });
-    await handler();
-    console.log(`Migration v${version} complete`);
+    logInfo(`Running migration v${version}: ${description}`, {
+      source: 'migrations',
+    });
+    try {
+      const output = await handler();
+      await dbMigrations.upsertOne(
+        {
+          version,
+        },
+        {
+          $set: {
+            status: 'completed',
```

Comment:
@omegascorp Migrations can also have a "description" field, which is very helpful to keep in the DB, alongside versions.